lib/fputs.js

```diff
@@ -0,0 +1,185 @@
+     *
+     * options:
+     *     writemode:   file open mode, default 'a'
+     *     writesize:   number of chars to write per chunk, default 100k
+     */
+    function Fputs( writable, opts ) {
+        if (!(this instanceof Fputs)) return new Fputs(writable, opts);
+
+        opts = opts || {};
+        if (typeof writable === 'string') {
+            // convert a string filename into a FileWriter writable
+            writable = new Fputs.FileWriter(writable, opts.writemode || 'a');
+        }
+        if (!writable) writable = process.stdout;
+
+        this.writable = writable;
+        this.writesize = opts.writesize || 102400;
+        this.highWaterMark = opts.highWaterMark || this.writesize;
+        this.datachunks = [];
+        this.unwrittenLength = 0;
+        this.writtenLength = 0;
+        this.resetCount = 0;
+        this._syncing = false;
+
+        this._error = null;
+        this.reportError = function(err) { if (!this._error) this._error = err; }
+        this.returnError = function() { var err = this._error; this._error = null; return err; }
+
+        // expose FileWriter.renameFile
+        // FIXME: set to a function that returns an error if renameFile not available
+        if (writable._isFileWriter) this.renameFile = writable.renameFile;
+    }
+
+    // export FileWriter on the Fputs class
+    Fputs.FileWriter = FileWriter;
+
+    // export renameFile as a QFputs class method as well
+    Fputs.renameFile = FileWriter.renameFile;
+
+    /**
+     * Append a newline terminated string to the fifo.
+     */
+    Fputs.prototype.fputs = function fputs( str ) {
+        if (typeof str !== 'string') str = "" + str;
+        this.write((str[str.length - 1] === "\n") ? str : str + "\n");
+    }
+
+    /**
+     * Write bulk data to the target.  Newline termination is not checked.
+     */
+    Fputs.prototype.write = function write( str, callback ) {
+        if (typeof str !== 'string') str = "" + str;
+
+        // merge writes into this.writesize sized data chunks
+        // it is assumed that all writes end with a newline (not checked)
+        var nchunks = this.datachunks.length;
+        if (nchunks > 0 && this.datachunks[nchunks-1].length + str.length <= this.writesize) {
+            // the last chunk has space for more
+            this.datachunks[nchunks-1] += str;
+        }
+        else {
+            // else start a new chunk
+            this.datachunks.push(str);
+        }
+
+        if (this.writtenLength && this.writtenLength >= this.unwrittenLength) {
+            // reset the lengths to avoid numeric overflow
+            this.writtenLength = this.unwrittenLength = 0;
+            this.resetCount += 1;
+        }
+        this.unwrittenLength += str.length;
+
+        if (!this._syncing) {
+            // if not currently syncing, start the sync thread
+            var self = this;
+            setTimeout(function(){ self._sync(); }, 1);
```

Comment:
I was allowing the 1ms for any additional lines that also arrive to be write-combined
and written together.  Without, bursts of writes will make two write per burst, not one.
I've thought of making the delay a bit longer (say 2-5ms) as the sync happens in
the background and is not synchronous to the call, but it would add to the overall latency.